finality-aleph/src/sync/task_queue.rs

```diff
@@ -0,0 +1,108 @@
+use std::{
+    cmp::Ordering,
+    collections::BinaryHeap,
+    fmt::{Debug, Formatter},
+};
+
+use log::warn;
+use tokio::time::{sleep, Duration, Instant};
+
+use crate::sync::LOG_TARGET;
+
+#[derive(Clone, Eq, PartialEq)]
+struct ScheduledTask<T: Eq> {
+    task: T,
+    scheduled_time: Instant,
+}
+
+impl<T: Eq> PartialOrd for ScheduledTask<T> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl<T: Eq> Ord for ScheduledTask<T> {
+    /// Compare tasks so that earlier times come first in a max-heap.
+    fn cmp(&self, other: &Self) -> Ordering {
+        other.scheduled_time.cmp(&self.scheduled_time)
+    }
+}
+
+#[derive(Clone, Default)]
+pub struct TaskQueue<T: Eq + PartialEq> {
```

Comment:
Code itself doesn't require `T` to be anyhow related with some abstract `task`. Perhaps add some requirement for it or name it like `ScheduledQueue`. Or, perhaps that `ScheduledTask` should be implementing that `waiting` logic directly and then that whole queue is unecessary and can be just switched with any ordered queue. Consider also that `pop` can actually `block` whole queue until head task is ready.